types/buildroot/buildroot.go

```diff
@@ -13,10 +13,8 @@ var arenaPool fastrlp.ArenaPool
 func CalculateReceiptsRoot(receipts []*types.Receipt) types.Hash {
 	ar := arenaPool.Get()
 
-	res := calculateRootWithRlp(len(receipts), func(i int) *fastrlp.Value {
```

Comment:
Why did we change this, wouldn't this do more memory allocations?

Reply:
A simple reason is that we need to add TransactionType if tx has.
According to [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718), TransactionType comes before TransactionData.

```
// Add transaction type before tx
TransactionType || rlp([nonce, gasPrice, gasLimit, ...])
```

MarshalRLPWith creates one RLP value. So if I encode TransactionType inside MarshalRLPWith, data will be like this:

```
// Add transaction type before tx 
rlp([TransactionType, nonce, gasPrice, gasLimit, ...])
```

So we need to encode Transaction first and then add TransactionType